File: akka-sample-cqrs-scala/src/main/scala/sample/cqrs/ShoppingCartServer.scala
package sample.cqrs

import scala.concurrent.duration._
import scala.util.Failure
import scala.util.Success

import akka.actor.typed.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.server.Route

class ShoppingCartServer(routes: Route, port: Int)(implicit system: ActorSystem[_]) {

  import system.executionContext

  def start(): Unit = {
    Http().newServerAt("localhost", port).bind(routes)
      .map(_.addToCoordinatedShutdown(3.seconds))
      .onComplete {
      case Success(binding) =>
        val address = binding.localAddress
        system.log.info("Shopping online at http://{}:{}/", address.getHostString, address.getPort)

      case Failure(ex) =>
        system.log.error("Failed to bind HTTP endpoint, terminating system", ex)
        system.terminate()
    }
  }

}

File: akka-sample-cqrs-java/src/main/java/sample/cqrs/ShoppingCartServer.java
package sample.cqrs;

import akka.actor.typed.ActorSystem;
import akka.actor.typed.javadsl.Adapter;
import akka.http.javadsl.Http;
import akka.http.javadsl.ServerBinding;
import akka.http.javadsl.server.Route;
import akka.stream.Materializer;

import java.net.InetSocketAddress;
import java.time.Duration;
import java.util.concurrent.CompletionStage;

class ShoppingCartServer {

  static void startHttpServer(Route route, int httpPort, ActorSystem<?> system) {
    // Akka HTTP still needs a classic ActorSystem to start
    akka.actor.ActorSystem classicSystem = Adapter.toClassic(system);
    final Http http = Http.get(classicSystem);
    final Materializer materializer = Materializer.matFromSystem(system);

    CompletionStage<ServerBinding> futureBinding =
      http.newServerAt("localhost", httpPort).bind(route);

    futureBinding
      .thenApply(binding -> binding.addToCoordinatedShutdown(Duration.ofSeconds(3), system))
      .whenComplete((binding, exception) -> {
        if (binding != null) {
          InetSocketAddress address = binding.localAddress();
          system.log().info("Server online at http://{}:{}/",
            address.getHostString(),
            address.getPort());
        } else {
          system.log().error("Failed to bind HTTP endpoint, terminating system", exception);
          system.terminate();
        }
      });
  }

}
